Save dependencies adjacency matrix in a png image

srmurguia opened the pull request:
For large projects with many module dependencies, the dependency graph
can be dificult to visualize. Now, a png image is generated with a
representation of the adjacency matrix.

The first row and first columns are headers, the colors there represent
the type of module:
  - Blue: Android app
  - Green: Android module
  - Yellow: Java module

If the cell in row r and column c is green, it means that the module
associated with that row has a dependency with the module associated
with that column. We can add different colors for different dependency
methods later (api, implementation, testImplementation, etc).

This image was created using these settings:

   "topologies": [
      {"type": "binary_tree", "seed": "2"}
   ],
  "dependencies": [
    {"from": "module1", "to": "module0"}
  ]

On a project with 20 Android and 50 pure modules

![dependencies](https://user-images.githubusercontent.com/29552205/38710336-bc63be02-3e74-11e8-9672-c4119a1dec5c.png)

srmurguia committed:
Save dependencies adjacency matrix in a png image

For large projects with many module dependencies, the dependency graph
can be dificult to visualize. Now, a png image is generated with a
representation of the adjacency matrix.

The first row and first columns are headers, the colors there represent
the type of module:
  - Blue: Android app
  - Green: Android module
  - Yellow: Java module

If the cell in row r and column c is green, it means that the module
associated with that row has a dependency with the module associated
with that column. We can add different colors for different dependency
methods later (api, implementation, testImplementation, etc).

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/models/ProjectBlueprint.kt:
```diff
@@ -79,6 +80,7 @@ class ProjectBlueprint(private val projectConfig: ProjectConfig) {
     fun saveDependencies() {
         val graphFileName = projectRoot.joinPath("dependencies.dot")
         File(graphFileName).printWriter().use { out -> out.print(dependenciesGraphString())}
+        DependencyImageGenerator().generate(this)
```
Please move the whole `saveDependencies()` method to `SourceModuleGenerator`, then inject the `DependencyImageGenerator()` into it, and use `FileWriter` to manipulate with files. And please rename `graphFileName` to `dependenciesGraphFileName` and keep it as a field in this class.
I'm sorry I missed you previous PR.

srmurguia replied:
I will move all functions related to the depency graph to its own generator, that will simplify things. Also, the files related to it (the .dot and the png image) aren't actually part of the modules (not even the project), they are only auxiliar files that help in visualizing the project's structure.

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/utils/DependencyImageGenerator.kt:
```diff
@@ -0,0 +1,130 @@
+     * generates an image that represents the dependency matrix of the project
+     */
+    fun generate(blueprint: ProjectBlueprint) {
+        val numModules = blueprint.moduleBlueprints.size + blueprint.androidModuleBlueprints.size
```
please make `allModuleBlueprints` in `ProjectBlueprint` public and use it here.

srmurguia replied:
This part no longer exists, but made allModuleBlueprints public as it is required for the new changes

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/utils/DependencyImageGenerator.kt:
```diff
@@ -0,0 +1,130 @@
+
+            // Dependencies
+            for (dependency in module.moduleDependencies) {
+                val split = DependencyValidator.ModuleSplit(dependency.name)
```
Please extract `ModuleSplit` into its own class, since it is needed in more that one place.

NikitaKozlov replied:
And `ANDROID_TYPE`, `MODULE_TYPE` constants as well.

NikitaKozlov replied:
BTW, ModuleSplit is `ConfigPOJO` dependent, so for I don't think it'll work for the new format at all.

NikitaKozlov replied:
As a solution you can always look for a module with a name, from there you can get its index and type. Then you don't need `SplitModule` and these indexes at all. Even more you can unite all this into one `forEach` stream with a few `maps`

srmurguia replied:
I was not thinking in the new format when I wrote this, we cannot use the name to get the type and the index from it since now the names can be anything. I will rewrite this part to work directly with the blueprint properties.

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/utils/DependencyImageGenerator.kt:
```diff
@@ -0,0 +1,130 @@
+            // Dependencies
+            for (dependency in module.moduleDependencies) {
+                val split = DependencyValidator.ModuleSplit(dependency.name)
+                val dependencyIndex = if (split.type == "module") {
```
Please use `MODULE_TYPE` constant here. Or even better add a method to split, that says `isAndroid` and another extension method `isNotAndroid` and use it here

srmurguia replied:
Same comment here, will rework to not depend on the name.

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/utils/DependencyImageGenerator.kt:
```diff
@@ -0,0 +1,130 @@
+
+        // Add Android modules headers
+        val numAndroid = blueprint.androidModuleBlueprints.size
+        for ((index, module) in blueprint.androidModuleBlueprints.withIndex()) {
```
Would be also nice if instead of this big `for` you can write `forEach` with mapping and everything.

srmurguia replied:
This was replaced to use blueprint.allDependencies.forEach

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/utils/DependencyImageGenerator.kt:
```diff
@@ -0,0 +1,130 @@
+
+        // Save to file
+        val imgPath = blueprint.projectRoot.joinPath("dependencies.png")
+        ImageIO.write(img,"png", File(imgPath))
```
Please make `ImageWriter` similar to `FileWriter`, and use it here and in `ImageGenerator`

srmurguia replied:
There are two writers now, with that, it is easier to create tests for DependencyGraphGeneration, thanks for suggesting it!

borisf commented on src/main/kotlin/com/google/androidstudiopoet/utils/DependencyImageGenerator.kt:
```diff
@@ -0,0 +1,130 @@
+
+        // Save to file
+        val imgPath = blueprint.projectRoot.joinPath("dependencies.png")
+        ImageIO.write(img,"png", File(imgPath))
```
@NikitaKozlov let's make sure that the review comments are aligned with the [extending AS guide](https://github.com/android/android-studio-poet/blob/master/extend/Extending.md); let me know what need to be updated :)

NikitaKozlov replied:
I'll make a PR for that, I have a small thing in mind

srmurguia committed:
Move methods used for dependency graph into DependencyGraphGenerator

This alligns to the extending AS poet guide, in particular, it separates
the generator from the blueprint.

srmurguia committed:
Simplify and add tests for DependencyGraphGenerator

Also, changed colors to match Gradle documentation.

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/generators/DependencyGraphGenerator.kt:
```diff
@@ -0,0 +1,166 @@
+            val JAVA_COLOR = Color(0xF7DB64)
+            val ERROR_COLOR = Color(0xDF0000)
+            // Colors match https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph
+            val dependencyMethodToColor = mapOf(
```
There might be a problem with flavours, but we'll tackle it later.

srmurguia replied:
Yes, different flavors can have different dependencies, we can either have all dependencies for all flavors in this image or generate one per variant. I think it would be better to have a single image because this image purpose is to have a summary of the dependencies.

NikitaKozlov commented on src/main/kotlin/com/google/androidstudiopoet/generators/DependencyGraphGenerator.kt:
```diff
@@ -0,0 +1,166 @@
+        val dependenciesGraphFileName = projectBlueprint.projectRoot.joinPath("dependencies.dot")
+        fileWriter.writeToFile(dependenciesGraphString(projectBlueprint), dependenciesGraphFileName)
+        println("Dependency graph written to $dependenciesGraphFileName")
+        DependencyImageGenerator(imageWriter).generate(projectBlueprint)
```
Can you please inject `DependencyImageGenerator` instead of creating it here?

srmurguia replied:
What do you mean by inject? Pass as parameter?

NikitaKozlov replied:
In the constructor :)

srmurguia replied:
Done, also, moved DependencyImageGenerator to its own file, that movement simplified the tests by now having 2 test files also (that depend in a common base)

srmurguia committed:
Move DependencyImageGenerator to it own file, with its own tests

This change simplifies tests for DependencyGrapGenerator and also adds a
test to check image content by comparing to a golden image.
